atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/ValidatingTransactionScopedCacheTest.java

```diff
@@ -0,0 +1,121 @@
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CellReference;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.TransactionFailedNonRetriableException;
+import com.palantir.common.streams.KeyedStream;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class ValidatingTransactionScopedCacheTest {
+    private static final TableReference TABLE = TableReference.createFromFullyQualifiedName("t.table");
+    private static final Cell CELL_1 = createCell(1);
+    private static final Cell CELL_2 = createCell(2);
+    private static final ImmutableSet<Cell> CELLS = ImmutableSet.of(CELL_1, CELL_2);
+    private static final CacheValue VALUE_1 = createValue(10);
+    private static final CacheValue VALUE_2 = createValue(20);
+    private static final ImmutableMap<Cell, byte[]> VALUES = ImmutableMap.<Cell, byte[]>builder()
+            .put(CELL_1, VALUE_1.value().get())
+            .put(CELL_2, VALUE_2.value().get())
+            .build();
+
+    private BiFunction<TableReference, Set<Cell>, ListenableFuture<Map<Cell, byte[]>>> valueLoader;
+
+    @Before
+    public void before() {
+        valueLoader = mock(BiFunction.class);
+    }
+
+    @Test
+    public void validationCausesAllCellsToBeReadRemotely() {
+        TransactionScopedCache delegate = TransactionScopedCacheImpl.create(snapshotWithSingleValue());
+        TransactionScopedCache validatingCache = new ValidatingTransactionScopedCache(delegate, () -> true);
+
+        when(valueLoader.apply(TABLE, CELLS)).thenReturn(remoteRead(CELLS));
+
+        validatingCache.get(TABLE, CELLS, valueLoader);
+        verify(valueLoader).apply(TABLE, CELLS);
+    }
+
+    @Test
+    public void validationDoesNotReadFromRemoteWhenItShouldNotValidate() {
+        TransactionScopedCache delegate = TransactionScopedCacheImpl.create(snapshotWithSingleValue());
+        TransactionScopedCache validatingCache = new ValidatingTransactionScopedCache(delegate, () -> false);
+
+        validatingCache.get(TABLE, ImmutableSet.of(CELL_1), valueLoader);
+        verify(valueLoader, never()).apply(any(), any());
+    }
+
```

Comment:
Add a test that we still read from remote when our cache does not have the entry

Reply:
That's actually already tested - `validationCausesAllCellsToBeReadRemotely` includes a cell that was not cached at the start too